parser/preinstrumentationtracing.go

```diff
@@ -0,0 +1,101 @@
+package parser
+
+import (
+	"github.com/dave/dst"
+	"github.com/dave/dst/dstutil"
+)
+
+// DetectTransactions analyzes the AST to identify and track transactions within function declarations.
+// It updates the transaction cache with function declarations and expressions related to transactions.
+func DetectTransactions(manager *InstrumentationManager, c *dstutil.Cursor) {
+	funcNode := c.Node()
+	if decl, ok := funcNode.(*dst.FuncDecl); ok {
+		manager.transactionCache.Functions[decl.Name.Name] = decl
+
+		var currentTransaction *dst.Ident
+		var recording bool
+		dstutil.Apply(decl.Body, func(c *dstutil.Cursor) bool {
+			node := c.Node()
+			switch stmt := node.(type) {
+			case *dst.AssignStmt:
+				for _, rhs := range stmt.Rhs {
+					callExpr, ok := rhs.(*dst.CallExpr)
+					if !ok {
+						continue
+					}
+					selExpr, ok := callExpr.Fun.(*dst.SelectorExpr)
+					if !ok {
+						continue
+					}
+					_, ok = selExpr.X.(*dst.Ident)
+					if ok && selExpr.Sel.Name == "StartTransaction" {
+						// Capture the transaction variable name
+						if len(stmt.Lhs) > 0 {
+							txnVar, ok := stmt.Lhs[0].(*dst.Ident)
+							if ok {
+								currentTransaction = txnVar
```

Comment:
Add nil check for transaction. If we have a valid current transaction do we still need recording var?